### ISSUE TITLE
azblob: unexpected fault

### ISSUE THREAD

**Issue by zalegrala** [Azure/azure-sdk-for-go#21555]
### Bug Report

- import path of package in question, e.g. `.../services/compute/mgmt/2018-06-01/compute`
```
	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
	"github.com/Azure/azure-sdk-for-go/sdk/azcore/streaming"
	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob"
	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/blob"
	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/bloberror"
	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/blockblob"
	"github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/container"
```
- SDK version 1.7.1
- output of `go version`: go1.21.0 running in github actions.

- What happened?

During a test run, we saw this once and are unable to reproduce it.
```
unexpected fault address 0x7f285372ff80
fatal error: fault
[signal SIGSEGV: segmentation violation code=0x1 addr=0x7f285372ff80 pc=0x4ad629]

goroutine 79 [running]:
runtime.throw({0x158cc47?, 0x0?})
	/opt/hostedtoolcache/go/1.21.0/x64/src/runtime/panic.go:1077 +0x5c fp=0xc00042f360 sp=0xc00042f330 pc=0x473ebc
runtime.sigpanic()
	/opt/hostedtoolcache/go/1.21.0/x64/src/runtime/signal_unix.go:875 +0x285 fp=0xc00042f3c0 sp=0xc00042f360 pc=0x48c785
runtime.memmove()
	/opt/hostedtoolcache/go/1.21.0/x64/src/runtime/memmove_amd64.s:339 +0x3c9 fp=0xc00042f3c8 sp=0xc00042f3c0 pc=0x4ad629
runtime.slicecopy(0xc0000c2000, 0x8000?, 0x7f2853728000, 0x4adb39?, 0x0?)
	/opt/hostedtoolcache/go/1.21.0/x64/src/runtime/slice.go:344 +0xb1 fp=0xc00042f410 sp=0xc00042f3c8 pc=0x48e151
bytes.(*Reader).Read(0xc0000ac000, {0xc0000c2000, 0x8000, 0x0?})
	/opt/hostedtoolcache/go/1.21.0/x64/src/bytes/reader.go:44 +0xed fp=0xc00042f470 sp=0xc00042f410 pc=0x5afd8d
github.com/Azure/azure-sdk-for-go/sdk/azcore/internal/exported.(*nopCloser).Read(0xc00004a020, {0xc0000c2000, 0x8000, 0x8000})
	<autogenerated>:1 +0x6d fp=0xc00042f4b8 sp=0xc00042f470 pc=0xa34d2d
github.com/Azure/azure-sdk-for-go/sdk/azcore/runtime.(*retryableRequestBody).Read(0xc00004a0b0, {0xc0000c2000, 0x8000, 0x8000})
	/home/runner/work/tempo/tempo/vendor/github.com/Azure/azure-sdk-for-go/sdk/azcore/runtime/policy_retry.go:225 +0x5c fp=0xc00042f500 sp=0xc00042f4b8 pc=0xa600dc
io.(*LimitedReader).Read(0xc0003d8108, {0xc0000c2000, 0x8000, 0x8000})
	/opt/hostedtoolcache/go/1.21.0/x64/src/io/io.go:480 +0xc6 fp=0xc00042f560 sp=0xc00042f500 pc=0x521226
io.copyBuffer({0x1764160, 0xc00004a1b0}, {0x1762d20, 0xc0003d8108}, {0x0, 0x0, 0x0})
	/opt/hostedtoolcache/go/1.21.0/x64/src/io/io.go:430 +0x27e fp=0xc00042f608 sp=0xc00042f560 pc=0x520f7e
io.Copy(...)
	/opt/hostedtoolcache/go/1.21.0/x64/src/io/io.go:389
net.genericReadFrom({0x17641e0, 0xc0001222f0}, {0x1762d20, 0xc0003d8108})
	/opt/hostedtoolcache/go/1.21.0/x64/src/net/net.go:[671](https://github.com/grafana/tempo/actions/runs/6182568748/job/16782571811#step:4:672) +0x8c fp=0xc00042f[678](https://github.com/grafana/tempo/actions/runs/6182568748/job/16782571811#step:4:679) sp=0xc00042f608 pc=0x84c3ac
net.(*TCPConn).readFrom(0xc0001222f0, {0x1762d20, 0xc0003d8108})
	/opt/hostedtoolcache/go/1.21.0/x64/src/net/tcpsock_posix.go:54 +0xc9 fp=0xc00042f6e0 sp=0xc00042f678 pc=0x857a89
net.(*TCPConn).ReadFrom(0xc0001222f0, {0x1762d20, 0xc0003d8108})
	/opt/hostedtoolcache/go/1.21.0/x64/src/net/tcpsock.go:130 +0x6c fp=0xc00042f760 sp=0xc00042f6e0 pc=0x854dec
io.copyBuffer({0x17641e0, 0xc0001222f0}, {0x1762d20, 0xc0003d8108}, {0x0, 0x0, 0x0})
	/opt/hostedtoolcache/go/1.21.0/x64/src/io/io.go:416 +0x1bf fp=0xc00042f808 sp=0xc00042f760 pc=0x520ebf
io.Copy(...)
	/opt/hostedtoolcache/go/1.21.0/x64/src/io/io.go:389
net/http.persistConnWriter.ReadFrom({0xc0000a3988?}, {0x1762d20, 0xc0003d8108})
	/opt/hostedtoolcache/go/1.21.0/x64/src/net/http/transport.go:1801 +0x95 fp=0xc00042f888 sp=0xc00042f808 pc=0xa13415
bufio.(*Writer).ReadFrom(0xc0003f0640, {0x1762d20, 0xc0003d8108})
	/opt/hostedtoolcache/go/1.21.0/x64/src/bufio/bufio.go:797 +0x28e fp=0xc00042f938 sp=0xc00042f888 pc=0x5c0fce
io.copyBuffer({0x1762e20, 0xc0003f0640}, {0x1762d20, 0xc0003d8108}, {0x0, 0x0, 0x0})
	/opt/hostedtoolcache/go/1.21.0/x64/src/io/io.go:416 +0x1bf fp=0xc00042f9e0 sp=0xc00042f938 pc=0x520ebf
io.Copy(...)
	/opt/hostedtoolcache/go/1.21.0/x64/src/io/io.go:389
net/http.(*transferWriter).doBodyCopy(0xc0001263c0, {0x1762e20, 0xc0003f0640}, {0x1762d20, 0xc0003d8108})
	/opt/hostedtoolcache/go/1.21.0/x64/src/net/http/transfer.go:412 +0x5d fp=0xc00042fa40 sp=0xc00042f9e0 pc=0x9fe43d
net/http.(*transferWriter).writeBody(0xc0001263c0, {0x1762e20, 0xc0003f0640})
	/opt/hostedtoolcache/go/1.21.0/x64/src/net/http/transfer.go:370 +0x79a fp=0xc00042fbb8 sp=0xc00042fa40 pc=0x9fdc3a
net/http.(*Request).write(0xc0000c0200, {0x1762e20, 0xc0003f0640}, 0x0, 0xc0000ac180, 0x0)
	/opt/hostedtoolcache/go/1.21.0/x64/src/net/http/request.go:717 +0x12ad fp=0xc00042fe18 sp=0xc00042fbb8 pc=0x9d[684](https://github.com/grafana/tempo/actions/runs/6182568748/job/16782571811#step:4:685)d
net/http.(*persistConn).writeLoop(0xc000176240)
	/opt/hostedtoolcache/go/1.21.0/x64/src/net/http/transport.go:2424 +0x313 fp=0xc00042ffc0 sp=0xc00042fe18 pc=0xa17cb3
net/http.(*Transport).dialConn.func6()
	/opt/hostedtoolcache/go/1.21.0/x64/src/net/http/transport.go:1777 +0x34 fp=0xc00042ffe0 sp=0xc00042ffc0 pc=0xa12b94
runtime.goexit()
	/opt/hostedtoolcache/go/1.21.0/x64/src/runtime/asm_amd64.s:1650 +0x1 fp=0xc00042ffe8 sp=0xc00042ffe0 pc=0x4ac481
created by net/http.(*Transport).dialConn in goroutine 58
	/opt/hostedtoolcache/go/1.21.0/x64/src/net/http/transport.go:1777 +0x256b
```

- What did you expect or want to happen?  Several of my team have attempted to reproduce this locally and have been unable, but I wanted to file an issue in case something stands out here.
- How can we reproduce it?  I am currently unable to reproduce this.
- Anything we should know about your environment.  This was a test running in Github actions.

**Comment by jhendrixMSFT** [comment]
I'm guessing this happens calling some API in `azblob` to download a blob.  Do you know which API?

**Comment by zalegrala** [comment]
We've seen this again in the same area of code, running in Github Actions and failing to reproduce locally.

**Comment by zalegrala** [comment]
Running the tests with `-count=1000` and commenting portions, it appears that this only shows up in `(*blockblog.Client).UploadStream()`.

**Comment by jhendrixMSFT** [comment]
Are the tests running on Linux (I assume yes)?  Are they talking to Azure or the storage emulator?

**Comment by jhendrixMSFT** [comment]
Also, are you using the default config values when calling `UploadStream()`?  If you have a link to the test that would also help.

**Comment by zalegrala** [comment]
The specific test is attempting to write a payload larger than the block size to force the rollover.  The test is here, which calls into our code a bit.
https://github.com/grafana/tempo/blob/b300780a636ec409c5d69168d6e3d38326dd63a0/tempodb/backend/azure/azure_test.go#L116-L133

Here is the `writer` where we make the actual call.
https://github.com/grafana/tempo/blob/b300780a636ec409c5d69168d6e3d38326dd63a0/tempodb/backend/azure/azure.go#L328-L338

I was able to reproduce this locally if I ran the test thousands of times.  We see this more commonly in github action, and I'm uncertain why.  I thought perhaps it had to do with resource contention.

**Comment by jhendrixMSFT** [comment]
Thanks, I can repro this now.  Investigating...

**Comment by jhendrixMSFT** [comment]
If you read the `Body` of the `*http.Request` in your [`http.HandlerFunc`](https://github.com/grafana/tempo/blob/b300780a636ec409c5d69168d6e3d38326dd63a0/tempodb/backend/azure/azure_test.go#L140-L148) the crash goes away.  I haven't worked out _why_ this is the case but it should at least get you unblocked.  Here's what I did in my repro.
```go
	fake := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if _, err := io.ReadAll(r.Body); err != nil {
			panic(err)
		}
		w.WriteHeader(http.StatusCreated)
	}))
```

**Comment by zalegrala** [comment]
That's good insight.  We're converting this SDK into a versioned choice that we can migrate to over time using config, so not blocked, but would like to migrate to the new SDK due to the deprecation on the older client we've been using.  Strange that it doesn't reproduce for us all the time.  Would love to hear thoughts when you discover why that might be the case.

**Comment by jhendrixMSFT** [comment]
In [`copyFromReader`](https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/storage/azblob/blockblob/chunkwriting.go#L33) we use memory-mapped files for temporary buffers when uploading blob chunks, and when this function returns (after all chunks have been uploaded), the deferred `buffers.Free()` cleans them up.  What I've noticed is that in the cases where the crash happens, either `buffers.Free()` is in progress or has already executed. i.e. the memory-mapped files have been deleted _before_ the transport attempts to read from them.

Debugging into the HTTP server, when the `http.HandlerFunc` _doesn't_ read the request's `Body` the server ends up going down [this](https://github.com/golang/go/blob/master/src/net/http/server.go#L2025-L2046) code path in order to clean up any active requests.  However, it appears that this _can_ happen _after_ our `buffers.Free()` has executed (or possibly simultaneously), which would explain the inconsistent repro (i.e. the clean-up of HTTP requests races with the clean-up of memory-mapped files).

I don't think this is something we can prevent on the SDK's side (well, not using memory-mapped files would solve it but that has significant perf implications).  Once `copyFromReader` returns, all of the chunks have been sent to the server and there's an implicit assumption that the server read the chunks (I don't see how this could ever _not_ be the case).

**Comment by LasseHels** [comment]
@jhendrixMSFT Good stuff! I spent a lot of time yesterday on this issue and was not able to identify the root cause. Reading your analysis was cathartic :+1:

**Comment by jhendrixMSFT** [comment]
Thanks!  It was an interesting issue to investigate.

Since we now know the root cause and how to resolve it, if there are no objections, I'm going to close this issue.

**Comment by zalegrala** [comment]
Thanks @jhendrixMSFT, I appreciate the investigation.
